rust/parquet_derive/src/parquet_field.rs

```diff
@@ -0,0 +1,942 @@
+            "NaiveDateTime" => Some(ThirdPartyType::ChronoNaiveDateTime),
+            "NaiveDate" => Some(ThirdPartyType::ChronoNaiveDate),
+            "Uuid" => Some(ThirdPartyType::Uuid),
+            _ => None,
+        };
+
+        Field {
+            ident: f
+                .ident
+                .clone()
+                .expect("we only support structs with named fields"),
+            ty,
+            is_a_byte_buf,
+            third_party_type,
+        }
+    }
+
+    /// Takes the parsed field of the struct and emits a valid
+    /// column writer snippet. Should match exactly what you
+    /// would write by hand.
+    ///
+    /// Can only generate writers for basic structs, for example:
+    ///
+    /// struct Record {
+    ///   a_bool: bool,
+    ///   maybe_a_bool: Option<bool>
+    /// }
+    ///
+    /// but not
+    ///
+    /// struct UnsupportedNestedRecord {
+    ///   a_property: bool,
+    ///   nested_record: Record
+    /// }
+    ///
+    /// because this parsing logic is not sophisticated enough for definition
+    /// levels beyond 2.
+    pub fn writer_snippet(&self) -> proc_macro2::TokenStream {
+        let ident = &self.ident;
+        let column_writer = self.ty.column_writer();
+
+        let vals_builder = match &self.ty {
+            Type::TypePath(_) => self.copied_direct_vals(),
+            Type::Option(ref first_type) => match **first_type {
+                Type::TypePath(_) => self.option_into_vals(),
+                Type::Reference(_, ref second_type) => match **second_type {
+                    Type::TypePath(_) => self.option_into_vals(),
+                    _ => unimplemented!("sorry charlie"),
+                },
+                ref f @ _ => unimplemented!("whoa: {:#?}", f),
+            },
+            Type::Reference(_, ref first_type) => match **first_type {
+                Type::TypePath(_) => self.copied_direct_vals(),
+                Type::Option(ref second_type) => match **second_type {
+                    Type::TypePath(_) => self.option_into_vals(),
+                    Type::Reference(_, ref second_type) => match **second_type {
+                        Type::TypePath(_) => self.option_into_vals(),
+                        _ => unimplemented!("sorry charlie"),
+                    },
+                    ref f @ _ => unimplemented!("whoa: {:#?}", f),
+                },
+                ref f @ _ => unimplemented!("whoa: {:#?}", f),
+            },
+            f @ _ => unimplemented!("don't support {:#?}", f),
+        };
+
+        let definition_levels = match &self.ty {
+            Type::TypePath(_) => None,
+            Type::Option(ref first_type) => match **first_type {
+                Type::TypePath(_) => Some(self.optional_definition_levels()),
+                Type::Option(_) => unimplemented!("nested options? that's weird"),
+                Type::Reference(_, ref second_type)
+                | Type::Vec(ref second_type)
```

Comment:
Should the case analysis for `definition_levels` match that of `vals_builder`? for instance, for the former, it is OK for `Option<Vec<T>>` but panics in the former. 

It seems we can use the same pattern matching structure to generate both `vals_builder` and `definition_levels`.